x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java

```diff
@@ -514,6 +529,81 @@ public void testLicenseUpdateFailureHandlerUpdate() throws Exception {
+            .put("xpack.security.enabled", false)
+            .put("path.home", createTempDir())
+            .build();
+        SettingsModule settingsModule = new SettingsModule(Settings.EMPTY);
+        ThreadPool threadPool = new TestThreadPool(getTestName());
+
+        try {
+            UsageService usageService = new UsageService();
+            Security security = new Security(settings, null);
+
+            appender.addExpectation(new MockLogAppender.SeenEventExpectation(
+                "Security rest wrapper", ActionModule.class.getName(), Level.DEBUG,
+                "Using REST wrapper from plugin org.elasticsearch.xpack.security.Security"
+            ));
+
+            ActionModule actionModule = new ActionModule(settingsModule.getSettings(),
+                TestIndexNameExpressionResolver.newInstance(threadPool.getThreadContext()),
+                settingsModule.getIndexScopedSettings(), settingsModule.getClusterSettings(), settingsModule.getSettingsFilter(),
+                threadPool, Arrays.asList(security), null, null, usageService, null);
+            actionModule.initRestHandlers(null);
+
+            appender.assertAllExpectationsMatched();
+        } finally {
+            threadPool.shutdown();
+            appender.stop();
+            Loggers.removeAppender(amLogger, appender);
+        }
+    }
+
+    public void test3rdPartyHandlerIsNotInstalled() {
+        Settings settings = Settings.builder()
+            .put("xpack.security.enabled", false)
+            .put("path.home", createTempDir())
+            .build();
+
+        SettingsModule settingsModule = new SettingsModule(Settings.EMPTY);
+        ThreadPool threadPool = new TestThreadPool(getTestName());
+
+        class FakeHandler implements RestHandler {
+            @Override
+            public void handleRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
+            }
+        }
+        ActionPlugin secPlugin = new ActionPlugin() {
+            @Override
+            public UnaryOperator<RestHandler> getRestHandlerWrapper(ThreadContext threadContext) {
+                return handler -> new FakeHandler();
+            }
+        };
+
+        try {
+            UsageService usageService = new UsageService();
+            Security security = new Security(settings, null);
+
+            Exception e = expectThrows(IllegalArgumentException.class, () ->
+                new ActionModule(settingsModule.getSettings(),
+                TestIndexNameExpressionResolver.newInstance(threadPool.getThreadContext()),
+                settingsModule.getIndexScopedSettings(), settingsModule.getClusterSettings(), settingsModule.getSettingsFilter(),
+                threadPool, Arrays.asList(security, secPlugin), null, null, usageService, null)
+            );
+            assertThat(e.getMessage(), is("The org.elasticsearch.xpack.security.SecurityTests$3 plugin tried to install a custom REST" +
```

Comment:
We shouldn't use an anonymous class for this. It means that the message is dependent on the synthentic name generated by the compiler, and that's too fragile to use.

Reply:
For the warning headers, we can as well wait for #71061 to be merged.